### ISSUE TITLE
Duplicate <meta name="viewport"> if the custom one is defined in _document.js

### ISSUE THREAD

**Issue by kachkaev** [zeit/next.js#6919]
# Bug report

I recently upgraded to the latest canary and noticed horizontal scrolling on some mobile devices. As it turned out, this was caused by #6754 (cc @developit).

Although the default new value of `viewport` is inserted only when there is no custom one, this logic did not work for me, because my custom `viewport` tag was in `_document.js` and not inside `<Head />`.

```
<meta name="viewport" content="width=device-width" />
```

Moving the tag into `<Head />` solved the issue, but it took a while to investigate.

## Expected behavior

It'd be great if Next.js searched for `<meta name="viewport" content="..." />` inside custom `_document.js` and either did not insert the default one or asked developers to move their tag into `<Head />`. WDYT folks?

## System information
 - Version of Next.js: 8.0.5-canary.17

**Comment by wth1999** [comment]
same issue on `8.0.4`
I was trying to set `viewport-fit=cover` in the viewport meta tag in `_document.js`, but it was overridden by the default tag.
I believe it will be a better approach if the viewport meta is already set in `_document.js`, the default tag shouldn't be added

**Comment by smackgg** [comment]
Anybody see this question？

**Comment by lisamartin00** [comment]
Our team ran into this problem as well and ended up moving our `<meta viewport...>` tag from `_document.js` and into `_app.js`.  In `_app.js`, we are importing `Head from 'next/head';` which seems to remove the duplicate, whereas `_document.js` is importing `{ Head } from 'next/document';`, which did not remove the duplicate.  
`_app.js` seems like an appropriate place to put a global `<meta viewport...>` tag since it's not SSR related anyway.

**Comment by cesarvarela** [comment]
So the approach at the moment is to use _app instead of _document?

**Comment by lisamartin00** [comment]
@cesarvarela  yes that is our approach at the moment and does seem to be working..but if anyone has any other suggestions or best practices, we'd love to hear them!

**Comment by MarchWorks** [comment]
@timneutkens should I make PR to rise a warning in production mode like for the title

**Comment by cheeki** [comment]
In my case, whenever i use "key" property on <meta name="viewport" ... />, unwanted duplicated meta tag appears like below. even in `_app.js`
```
<meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1" class="next-head">
```

**Comment by MarchWorks** [comment]
@cheeki I made a PR to fix that but logically there is no need to use the `key` property on meta tags

**Comment by vzaidman** [comment]
key is needed if you have a child component rewriting the father's component "head".

**Comment by vzaidman** [comment]
_app:

`<meta key="meta_description" name="description" content="general description"/>`

ChildPage:

`<meta key="meta_description" name="description" content="page description"/>`

**Comment by MarchWorks** [comment]
I would like to confirm the **Expected behavior** so that I can solve this one as soon as possible

- If there is a `viewport` in `_document` don't add the default one
- If there is a `viewport` in `_document` and another in a `page` and they don't have the same `key` value add both of them
- If there is a `viewport` in `_document` and another in a `page` and they do have the same `key` value add the one in the `page`

**Comment by vzaidman** [comment]
I think it's correct.
but also notice `viewport` might be in `_app`

**Comment by MarchWorks** [comment]
> If there is a `viewport` in `_document` and another in a `page` and they do have the same `key` value add the one in the `page`

@vzaidman what if they don't a key value? I am considering it to be the same as the above

**Comment by MarchWorks** [comment]
@cheeki 

> In my case, whenever i use "key" property on <meta name="viewport" ... />, unwanted duplicated meta tag appears like below. even in `_app.js`
> 
> ```
> <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1" class="next-head">
> ```

I think that it's the right behavior check @vzaidman https://github.com/zeit/next.js/issues/6919#issuecomment-487374749 so if you want ro overwrite the default meta you shouldn't use the key

**Comment by MarchWorks** [comment]
@ijjk can you read the last 4 comments and confirm the desired behavior so that I can add the check here
https://github.com/zeit/next.js/blob/064be18039f9786976eb3d6dd5dcca3369741fc1/packages/next/pages/_document.tsx#L179-L180

**Comment by vzaidman** [comment]
The expected behaviour:
* every meta component without "key" is added to the head.
* every meta component with key overrides all the other meta components with the same key so only one will exist. the overrides will happen in the following order:
  1. page
  1. _app
  1. _document

**Comment by Janpot** [comment]
I think I'm seeing the same issue with 
```jsx
<meta charSet='utf-8' />
```

**Comment by Timer** [comment]
Hi! After careful review we've decided this is the intended behavior. Head elements that are expected to be deduped should be added using `next/head`.
The reason this surfaced is because we started defaulting a more commonly used `<meta>` tag -- not that something is working unexpectedly.

---

The correct solution in this scenario is to move to using `_app` over `_document`.

**Comment by Soundvessel** [comment]
It might help to mention this in the documentation https://nextjs.org/docs#custom-document

**Comment by Timer** [comment]
Feel free to send a PR, @Soundvessel!

**Comment by samilyak** [comment]
> Head elements that are expected to be deduped should be added using next/head

A downside of this approach is whenever you need some npm package to build global `<meta>`s which are always the same (e.g. `mobile-detect` package to add `<meta name="viewport">` conditionally) you're forced to include it on both server and client side.
Even though the package is actually needed only once in `_document` on server side (as this type of `<meta>`s don't need to be re-rendered in `_app`).

**Comment by stratboy** [comment]
I have the same issue. Quite crazy it's still not really solved.

**Comment by digitaltim-de** [comment]
Same problem... and can't fix it.

**Comment by stratboy** [comment]
> Same problem... and can't fix it.

Hi, for now, I used _app.js and that solved the issue. Obviously, it's not exactly how it should be...

**Comment by muratozgul** [comment]
`<meta charSet="utf-8" />`
or
` <meta key="viewport" name="viewport" content="initial-scale=1.0, width=device-width" />`
should be direct child of `Head`. I was returning many tags within a react fragment and I was getting duplicates as well.
I inspected [PR#7155 ](https://github.com/vercel/next.js/pull/7155/files) and then moved the meta tags under `Head` and fixed it.

For example:
```
const CustomHead = ({ children }: CustomHeadProps) => {
  return (
    <Head>
      <meta charSet="utf-8" />
      <meta key="viewport" name="viewport" content="initial-scale=1.0, width=device-width" />
      <CommonTags />
      <SEOTags />
      <OpenGraphTags />
      <CompatibilityTags />
      <link rel="icon" href="/favicon.ico" />
      {children}
    </Head>
  );
};
```

**Comment by Toxiapo** [comment]
I was getting a duplicate `<link" rel='icon' href='/favicon.ico' /`, adding a "key" to it solved the problem.

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
